[eigen3] Update to 3.4.0

spinicist opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
  Updates Eigen to 3.4.0

- #### Which triplets are supported/not supported? Have you updated the [CI baseline]
  all

- #### Does your PR follow the [maintainer guide]
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

I think this is good to go but have opened it as a draft as it is my first time.

Commit:
[eigen3] Update to 3.4.0

Neumann-A commented on ports/eigen3/disable_pkgconfig_absolute_path_check.patch:
```diff
@@ -1,18 +0,0 @@
-     CACHE STRING "The directory relative to CMAKE_PREFIX_PATH where eigen3.pc is installed"
-     )
- 
-+if(0)
- foreach(var INCLUDE_INSTALL_DIR CMAKEPACKAGE_INSTALL_DIR PKGCONFIG_INSTALL_DIR)
-   if(IS_ABSOLUTE "${${var}}")
-     message(FATAL_ERROR "${var} must be relative to CMAKE_PREFIX_PATH. Got: ${${var}}")
-   endif()
- endforeach()
-+endif()
```
why? None of those variables should be absolute.

Neumann-A replied:
Nevermind you removed the patch ;)

spinicist replied:
Correct - vcpkg failed to compile Eigen on my local machine with the patch in place, so I removed it.

autoantwort commented:
You can download the error logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=58025&view=artifacts&pathAsName=false&type=publishedArtifacts). If you hover over the items, there are tree dots on the right that you can click

spinicist commented:
> You can download the error logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=58025&view=artifacts&pathAsName=false&type=publishedArtifacts). If you hover over the items, there are tree dots on the right that you can click

Okay, I took a look at the osx logs and for `theia` and `openmvg` these look like genuine regressions to me - they are build failures that reference Eigen. How does the chain of responsibility work here with `vcpkg`? I've never used those ports so I'm apprehensive about digging about in them.

(An aside: given how low-level Eigen is I'm actually impressed by how few breakages there are here. 3.3 -> 3.4 had some pretty big changes)

Neumann-A commented:
> How does the chain of responsibility work here with vcpkg?

In generell, for a PR to be merge able CI needs to be green, so all ports which are not green needs to be fixed unless it is a baseline regressions like paraview on osx. (and paraview/vtk works on windows and linux so there is no issue there with eigen3) Sometimes it is just enough to bump the dependent port to a newer version or just report the failure upstream and hopefully they fix it in an timely manner so that the patch can easily be integrated here.

strega-nil-ms commented:
/azp run

NancyLi1013 commented on ports/eigen3/vcpkg.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "eigen3",
+  "version-string": "3.4.0",
```
```suggestion
  "version-semver": "3.4.0",
```
This should be [version-semver](https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#version-semver).

NancyLi1013 commented on versions/e-/eigen3.json:
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "51a0bcbf78726733fa51ad1aa15efbc2f55a499c",
```
```suggestion
      "git-tree": "471f99d976ff3d1abd774c7d025c9f78dd4e4639",
```
The value for git-tree has changed now. So you need to update it.

Or you can run ` .\vcpkg x-add-version --overwrite eigen3` to update this.

aluaces commented:
@spinicist, would you mind to update your branch with those requested changes?  Otherwise the process is blocked.  Thanks!

NancyLi1013 commented:
> @spinicist, would you mind to update your branch with those requested changes? Otherwise the process is blocked. Thanks!

Thanks for your reminder. Since there is no response for a long time. I updated the changes just now.

spinicist commented:
Hello,

Thanks for picking this up. I have been on leave and have had a large amount to catch up on.

If I interpret the current build logs correctly, the problems with the regressions in other projects seem to have resolved?

NancyLi1013 commented:
> Hello,
> 
> Thanks for picking this up. I have been on leave and have had a large amount to catch up on.
> 
> If I interpret the current build logs correctly, the problems with the regressions in other projects seem to have resolved?

Yes, there is nothing else needs to be done now. Just wait for the CI test result. Thanks for your PR @spinicist.

spinicist commented:
Looks like those regressions in other projects are still there after all. Should I start reaching out to those projects to see if they have fixes that can be incorporated?

NancyLi1013 commented:
Except for `pangolin`, other ports seem to be caused by the update. Could you please look into them?

NancyLi1013 commented:
The failures caused by `pangolin `will be fixed by PR #20006.

aluaces commented:
I have taken a look and there are some errors that is not clear is they are eigen's fault or not.  For example, I have submitted a question to eigen mailing list in order to know what the solution is for the one affecting openmvg.

Nevertheless, I find a bit difficult to point to the exact failing location of the errors in the automated builds above: if I click on each one I am redirected to an error summary, but it belongs to the whole process and doesn't detail the exact lines and snippets of code failing.  Is there a way of seeing that instead of compiling it myself and watching the build logs?

autoantwort commented:
You can get the build logs: Click Below on *Details* -> next Page on *View more details on Azure Pipelines* -> then on *2 artifacts produced* -> search the entry in the list -> hover over it -> click on the tree dots on the right -> click -> download

aluaces commented:
I have submitted a fix for the openmvg issue in https://github.com/microsoft/vcpkg/pull/20056

NancyLi1013 commented:
@aluaces 

Could you please resolve the conflicts?

Wongboo commented:
@aluaces , looks as if those diffs are not complicated, or may be I can offer help

aluaces commented:
> @aluaces , looks as if those diffs are not complicated, or may be I can offer help

Yes, I agree.  It was just a lack of time, but if you feel like it, be my guest.

Wongboo commented:
@aluaces , looks like I can't fork a forked repo or try to fetch upstream for the repo without write access (maybe?), where (not how) can I resolve those conflicts?

aluaces commented:
> @aluaces , looks like I can't fork a forked repo or try to fetch upstream for the repo without write access (maybe?), where (not how) can I resolve those conflicts?

I don't know.  I'm in the same boat as you, the PR is not mine so I don't know how to modify it.

spinicist commented:
Can I help?

Apologies for going quiet, my day job has been extremely busy.

ras0219-msft commented:
I've fixed the merge conflicts and updated the homepage to https; this LGTM once the CI is green.

Thanks for the PR!

cenit commented:
this patch solves the issue with theia (or better, with the akaze third party lib that it includes...).
For shogun, i still have to start the work

```
--- a/libraries/akaze/src/nldiffusion_functions.cpp
+++ b/libraries/akaze/src/nldiffusion_functions.cpp
@@ -222,8 +222,8 @@ void halfsample_image(const RowMatrixXf& src, RowMatrixXf& dst) {
 
   // Do the whole resampling in one pass by using neighboring values. First, we
   // compute the borders.
-  const double x_kernel_size = static_cast<double>(src.cols()) / dst.cols();
-  const double y_kernel_size = static_cast<double>(src.rows()) / dst.rows();
+  const int x_kernel_size = static_cast<double>(src.cols()) / dst.cols();
+  const int y_kernel_size = static_cast<double>(src.rows()) / dst.rows();
 
   // Do simple linear interpolation.
   if (x_kernel_size == 2 && y_kernel_size == 2) {
@@ -237,8 +237,8 @@ void halfsample_image(const RowMatrixXf& src, RowMatrixXf& dst) {
     return;
   }
 
-  const double x_kernel_clamped_size = static_cast<int>(ceil(x_kernel_size));
-  const double y_kernel_clamped_size = static_cast<int>(ceil(y_kernel_size));
+  const int x_kernel_clamped_size = static_cast<int>(ceil(static_cast<double>(src.cols()) / dst.cols()));
+  const int y_kernel_clamped_size = static_cast<int>(ceil(static_cast<double>(src.rows()) / dst.rows()));
 
   // Set up precomputed factor matrices.
   Eigen::RowVectorXf x_kernel_mul(static_cast<int>(x_kernel_clamped_size)),

```

cenit commented:
upstream open issue with shogun

https://github.com/shogun-toolbox/shogun/issues/4870

spinicist commented:
> (ported to https://gitlab.com/libeigen/eigen/-/issues/1664 but without any other comments added unfortunately)

An FYI that Eigen had problems with their repo at the weekend. Currently no issues that were opened before last weekend have any comments. Hopefully Gitlab will be able to restore them in the near future.

Looking at the Bugzilla thread, it looks like the fix should be straightforward - swapping to the correct `.head()` or `.tail()` method instead of `.block`?

cenit commented:
> > (ported to https://gitlab.com/libeigen/eigen/-/issues/1664 but without any other comments added unfortunately)
> 
> An FYI that Eigen had problems with their repo at the weekend. Currently no issues that were opened before last weekend have any comments. Hopefully Gitlab will be able to restore them in the near future.
> 
> Looking at the Bugzilla thread, it looks like the fix should be straightforward - swapping to the correct `.head()` or `.tail()` method instead of `.block`?

might be. 
In the meantime, feel free to add the theia patch to fix that port!

ras0219-msft committed:
[theia] Update to support Eigen 3.4

Neumann-A commented:
> In the theia patch, can we just remove the double casting entirely?

I think the idea is to keep it in sync with the clamped calculation where the cast to double is required.

cenit commented:
why is theia completely skipped in CI?

smancill commented:
> let me have some more days to work on shogun, before removing the port...

This patch should fix shogun 6.1.4

https://github.com/NixOS/nixpkgs/pull/145161/files#diff-ec23d000d3ede3ea11db8129a04b33d1a924523cf2eb52fa247f2a394805971d

aminya commented:
Is there an update on this? Eigen 3.4 was released 3 months ago, and it adds necessary patches for critical bugs [like this](https://gitlab.com/libeigen/eigen/-/issues/2137)

spinicist commented:
> Is there an update on this? Eigen 3.4 was released 3 months ago, and it adds necessary patches for critical bugs [like this](https://gitlab.com/libeigen/eigen/-/issues/2137)

Apologies - after submitting the original PR my own motivation to get it accepted has declined somewhat, as I'm actually using Eigen 3.4 in vcpkg already (via the branch for this PR).

The issue is there are a couple of other ports that depend on Eigen which do not compile with 3.4. There is a patch to fix one of them above (shogun). I believe after that a similar patch is required for theia, but I know nothing about theia.

aminya commented:
How can I do that in my manifest? Should I change my whole vcpkg repository? 
I wish vcpkg would use fixed versions, so every other package did not break. It is like a loophole. The developers want to try Eigen 3.40, so they can fix their code. But because vcpkg does not ship it, because the developers have not fixed their code yet.

aluaces commented:
@aminya , you just make a branch with your changes or event clone this branch.

aminya commented:
@spinicist 
Could you please resolve the conflicts so we can see what the state of the CI is?

ras0219-msft commented:
>How can I do that in my manifest? Should I change my whole vcpkg repository?

You can use overlay ports[1] or registries[2] to replace just eigen with the copy from this PR.

[1] https://github.com/microsoft/vcpkg/blob/master/docs/users/manifests.md#vcpkg_overlay_ports and https://github.com/microsoft/vcpkg/blob/master/docs/specifications/ports-overlay.md
[2] https://github.com/microsoft/vcpkg/blob/master/docs/users/registries.md

LilyWangLL commented:
Ping @spinicist for response.

cenit commented:
here it is also a patch for shogun and eigen 3.4
(note that the openmvg patch i posted above is still not merged in this PR, while the one for theia was merged by @ras0219-msft )

```
--- a/src/shogun/machine/gp/MultiLaplaceInferenceMethod.cpp
+++ b/src/shogun/machine/gp/MultiLaplaceInferenceMethod.cpp
@@ -84,9 +84,9 @@ class CMultiPsiLine : public func_base
 		float64_t result=0;
 		for(index_t bl=0; bl<C; bl++)
 		{
-			eigen_f.block(bl*n,0,n,1)=K*alpha->block(bl*n,0,n,1)*CMath::exp(log_scale*2.0);
-			result+=alpha->block(bl*n,0,n,1).dot(eigen_f.block(bl*n,0,n,1))/2.0;
-			eigen_f.block(bl*n,0,n,1)+=eigen_m;
+			eigen_f.segment(bl*n,n)=K*alpha->segment(bl*n,n)*CMath::exp(log_scale*2.0);
+			result+=alpha->segment(bl*n,n).dot(eigen_f.segment(bl*n,n))/2.0;
+			eigen_f.segment(bl*n,n)+=eigen_m;
 		}
 
 		// get first and second derivatives of log likelihood
@@ -272,7 +272,7 @@ void CMultiLaplaceInferenceMethod::update_alpha()
 	{
 		Map<VectorXd> alpha(m_alpha.vector, m_alpha.vlen);
 		for(index_t bl=0; bl<C; bl++)
-			eigen_mu.block(bl*n,0,n,1)=eigen_ktrtr*CMath::exp(m_log_scale*2.0)*alpha.block(bl*n,0,n,1);
+			eigen_mu.segment(bl*n,n)=eigen_ktrtr*CMath::exp(m_log_scale*2.0)*alpha.segment(bl*n,n);
 
 		//alpha'*(f-m)/2.0
 		Psi_New=alpha.dot(eigen_mu)/2.0;
@@ -316,7 +316,7 @@ void CMultiLaplaceInferenceMethod::update_alpha()
 
 		for(index_t bl=0; bl<C; bl++)
 		{
-			VectorXd eigen_sD=eigen_dpi.block(bl*n,0,n,1).cwiseSqrt();
+			VectorXd eigen_sD=eigen_dpi.segment(bl*n,n).cwiseSqrt();
 			LLT<MatrixXd> chol_tmp((eigen_sD*eigen_sD.transpose()).cwiseProduct(eigen_ktrtr*CMath::exp(m_log_scale*2.0))+
 				MatrixXd::Identity(m_ktrtr.num_rows, m_ktrtr.num_cols));
 			MatrixXd eigen_L_tmp=chol_tmp.matrixU();
@@ -341,11 +341,11 @@ void CMultiLaplaceInferenceMethod::update_alpha()
 		VectorXd tmp2=m_tmp.array().rowwise().sum();
 
 		for(index_t bl=0; bl<C; bl++)
-			eigen_b.block(bl*n,0,n,1)+=eigen_dpi.block(bl*n,0,n,1).cwiseProduct(eigen_mu.block(bl*n,0,n,1)-eigen_mean_bl-tmp2);
+			eigen_b.segment(bl*n,n)+=eigen_dpi.segment(bl*n,n).cwiseProduct(eigen_mu.segment(bl*n,n)-eigen_mean_bl-tmp2);
 
 		Map<VectorXd> &eigen_c=eigen_W;
 		for(index_t bl=0; bl<C; bl++)
-			eigen_c.block(bl*n,0,n,1)=eigen_E.block(0,bl*n,n,n)*(eigen_ktrtr*CMath::exp(m_log_scale*2.0)*eigen_b.block(bl*n,0,n,1));
+			eigen_c.segment(bl*n,n)=eigen_E.block(0,bl*n,n,n)*(eigen_ktrtr*CMath::exp(m_log_scale*2.0)*eigen_b.segment(bl*n,n));
 
 		Map<MatrixXd> c_tmp(eigen_c.data(),n,C);
 
@@ -409,7 +409,7 @@ float64_t CMultiLaplaceInferenceMethod::get_derivative_helper(SGMatrix<float64_t
 	{
 		result+=((eigen_E.block(0,bl*n,n,n)-eigen_U.block(0,bl*n,n,n).transpose()*eigen_U.block(0,bl*n,n,n)).array()
 			*eigen_dK.array()).sum();
-		result-=(eigen_dK*eigen_alpha.block(bl*n,0,n,1)).dot(eigen_alpha.block(bl*n,0,n,1));
+		result-=(eigen_dK*eigen_alpha.segment(bl*n,n)).dot(eigen_alpha.segment(bl*n,n));
 	}
 
 	return result/2.0;
@@ -489,7 +489,7 @@ SGVector<float64_t> CMultiLaplaceInferenceMethod::get_derivative_wrt_mean(
 		result[i]=0;
 		//currently only compute the explicit term
 		for(index_t bl=0; bl<C; bl++)
-			result[i]-=eigen_alpha.block(bl*n,0,n,1).dot(eigen_dmu);
+			result[i]-=eigen_alpha.segment(bl*n,n).dot(eigen_dmu);
 	}
 
 	return result;
```

cenit commented:
@ras0219-msft since you have power to push here, can you merge with master and add my patches, so that we can see CI status at the end?

spinicist commented:
I have some time this week.

I have rebased from master and squashed the commits relating solely to Eigen. Apologies if this was the wrong thing to do.

@cenit If you fork my fork of vcpkg, do you want to add your patches as a PR to the eigen3.4, so that you get credit for them in the history?

cenit commented:
i could do that but i might not have time this week. if you have time please go on and apply them

spinicist committed:
[shogun/openmvg] Patches to support Eigen 3.4 from @cenit

spinicist commented:
Okay, fingers crossed I have incorporated the patches from @cenit correctly.

@LilyWangLL the Github actions bot still shows requests that I think have been met now?

cenit commented:
you forgot to bump port-version after recent patches

cenit commented:
after that, you will still need a manual approval for ci tests since you are a first time contributor

spinicist commented:
Okay, I have bumped the port-version for openmvg and shogun. Anything else?

cenit commented:
in the meantime openmvg was updated, so you have to merge conflicts...
sorry this road is so long, but eigen is a complicated port!

spinicist committed:
Commit result of x-add-version

spinicist commented:
I have re-applied the changes on top of the OpenMVG updates, assuming that the patch is still required.

Eigen itself was very straightforward to update. It's the dependency ports that are taking the time!

cenit commented:
> I have re-applied the changes on top of the OpenMVG updates, assuming that the patch is still required.
> 
> Eigen itself was very straightforward to update. It's the dependency ports that are taking the time!

complicate from that point of view ;)

spinicist commented:
I think this is getting close but the CI is confusing me.

1 - I think MacOS compiled everything, but then failed at the final step?
2 - Linux failed, I think because of a missing new-line at the end of the patch. I have added a new-line and updated.
3 - Windows failed because of an apparent regression in sophus, which this PR does not touch and I know nothing about.

I'd very much welcome some help / advice here, for MacOS and Windows I have no clue why the CI failed!

cenit commented:
linux and osx are fine now
i will try to understand problems with sophus asap

aminya commented:
The error log gives no information why the Sophus build fails! 

[install-x64-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/7842848/install-x64-windows-dbg-out.log)

spinicist commented:
@aminya Should this be done within the `portfile.cmake` or as a patch to apply to `CMakeLists.txt`? There is already a `fix_cmakelists.patch` I could edit?

spinicist commented:
Okay, CI is running again...

aminya commented:
It seems the patch you created is incorrect at line 18

spinicist commented:
Apologies, I tried hand-crafting the patch and made an error.

A new version with the output from diff is now in CI.

spinicist commented:
Okay, I cannot see what is wrong with the patch now, any ideas?

cenit commented:
without looking at the errors, i can suggest you to check end of line (use LF, not CRLF)

spinicist commented:
> without looking at the errors, i can suggest you to check end of line (use LF, not CRLF)

I'm working on a Mac, not Windows, so I don't think that's the problem. I downloaded it from Github, on another Mac, and ran file:

```
>file fix_cmakelists.patch 
fix_cmakelists.patch: unified diff output text, ASCII text
```

Looks correct I think?

It's complaining that line 19 is invalid, but there are only 18 in the file. So it could be a missing EOL before EOF, but the previous patch looks to be the same to me?

craigscott-crascit commented on ports/sophus/fix_cmakelists.patch:
```diff
@@ -1,12 +1,18 @@
  LIST( APPEND Sophus_INCLUDE_DIR ${EIGEN3_INCLUDE_DIR} )
-
+-
++IF(MSVC)
++  add_definition(/bigobj)
++ELSE()
++  add_definition(-Wa,-mbig-obj)
++ENDIF()
```
I'm a bit surprised this works. There is no standard CMake command called `add_definition()`. Presumably, you meant `add_definitions()`, but that would still be the wrong command to call here (but I see that higher up in the `CMakeLists.txt` file, it already misuses `add_definitions()` to add arbitrary compiler flags). The `add_compile_options()` command (available since CMake 2.8.12) would normally be what you would use for adding arbitrary compiler flags.

The `ELSE()` clause also seems overly generic. The `-Wa,-mbig-obj` options will be used for every toolchain except MSVC. Shouldn't this be added only for those toolchains we know support it? Look higher up in the `CMakeLists.txt` file and you'll see other flags being added only when the compiler ID matches `Clang` or `GNU`, for example.

spinicist replied:
Hello,

I merely copied @aminya 's suggestion above. What exactly would you recommend instead?

...and any idea why the patch appears to be failing overall? I can't see what is wrong but I'm not an expert in patch formatting.

aminya replied:
> The add_compile_options() command (available since CMake 2.8.12) would normally be what you would use for adding arbitrary compiler flags.

Yeah, please replace it with the standard `add_compile_options`.  I just used GitHub comments, and I didn't have an editor open in front of me to test it. 

> ...and any idea why the patch appears to be failing overall? I can't see what is wrong but I'm not an expert in patch formatting.

Have you tried creating the patch using Git itself?

spinicist replied:
> Have you tried creating the patch with git itself?

Yes, this is how I made the latest version of the patch.

aminya replied:
@spinicist
Hmm. That's strange then. Let's wait for #22792. Hopefully, it fixes the issue.

aminya commented on ports/sophus/vcpkg.json:
```diff
@@ -1,7 +1,7 @@
   "name": "sophus",
   "version": "1.0.0",
-  "port-version": 7,
+  "port-version": 8,
```
Why is vcpkg using a Sophus form 2017?! this is very outdated. This outdated package should not hold everyone else from using Eigen 3.4.0

https://github.com/strasdat/Sophus/compare/v1.0.0...master

aminya replied:
I made a PR to update this 
https://github.com/microsoft/vcpkg/pull/22792

aminya replied:
This is now merged! 🎉 @spinicist could you rebase?

aminya commented:
@spinicist could you rebase with the PR I made for sophus? I think you can remove the patch, and it should work. Now, sophus addds `/bigobj` in [their CMake file](https://github.com/strasdat/Sophus/blob/49a7e1286910019f74fb4f0bb3e213c909f8e1b7/CMakeLists.txt#L28).

aminya commented:
Can someone with access rebase this branch? @ras0219-msft @spinicist

spinicist commented:
> Can someone with access rebase this branch? @ras0219-msft @spinicist

It now needs more than rebase - the Sophus patches will need removing. I will try to find time but I have a lot on.

aminya commented:
> > Can someone with access rebase this branch? @ras0219-msft @spinicist
> 
> It now needs more than rebase - the Sophus patches will need removing. I will try to find time but I have a lot on.

@spinicist I made a PR. Here you go: https://github.com/spinicist/vcpkg/pull/1

spinicist commented:
I have updated the PR to remove the sophus patches. A local vcpkg install (on Mac) of sophus now works.

I am completely confused by the fresh set of failures in the CI. Linux and Windows x64 static pass cleanly.

x86 Windows complains that the shogun version is wrong - but several other pipelines say it is fine?
x64 Windows complains rtabmap didn't compile but I can't see why?
The OSX failure I can't begin to understand - looks like the CI as whole failed?

aminya commented:
rtabmap error is again because of bigobj

```
D:\buildtrees\rtabmap\src\fde13e9126-d5e18ef027.clean\corelib\src\util3d_registration.cpp : fatal error C1128: number of sections exceeded object file format limit: compile with /bigobj
ninja: build stopped: subcommand failed.
```

Adding this to rtabmap should fix the issue:
```
if(MSVC)
  add_compile_options(/bigobj)
endif()
```

aminya committed:
[rtabmap] add bigobj

aminya commented:
@spinicist I fixed the issues in this pull request:
https://github.com/spinicist/vcpkg/pull/1

aminya committed:
x-add-version

spinicist commented:
Thanks @aminya CI is running again now

aminya commented:
> Thanks @aminya CI is running again now

Whoohoo. 🎉 Everything passed. Should ask the maintainers to merge it.

cenit commented:
@LilyWangLL PR is ready :)

LilyWangLL commented:
Thanks for everyone' work on this PR. I think the fix for ports which depend on eigen3 should synced to their upstream.

JackBoosY commented on ports/eigen3/vcpkg.json:
```diff
@@ -1,7 +1,16 @@
   "name": "eigen3",
-  "version-string": "3.3.9",
-  "port-version": 1,
+  "version-semver": "3.4.0",
```
Any upstream evidence that they are using semver?

Neumann-A replied:
They probably are. But I think we should always just use version. @strega-nil replaced `version-semver` in the qt PR with just `version`.

spinicist replied:
I only changed to `version-semver` because the Github bot requested it. I will accept the change back to `version` - perhaps the Bot can be updated?

Neumann-A commented on ports/eigen3/vcpkg.json:
```diff
@@ -1,7 +1,16 @@
   "name": "eigen3",
-  "version-string": "3.3.9",
-  "port-version": 1,
+  "version-semver": "3.4.0",
```
```suggestion
  "version": "3.4.0",
```

spinicist committed:
Update ports/eigen3/vcpkg.json

Co-authored-by: Alexander Neumann <30894796+Neumann-A@users.noreply.github.com>

aminya commented:
@spinicist You have to run this again. I wish the reviewers could set up a bot to do this automatically when someone uses GitHub to commit a suggested change.

```
./vcpkg x-add-version --overwrite-version --all
```

JackBoosY commented:
@aminya Bot may not have permission to push to contributor's branch.

spinicist committed:
Run x-add-version

spinicist commented:
Okay, CI is running again...

spinicist commented:
I think this is good to go? Finally? 🤞

aminya commented:
> I think this is good to go? Finally? 🤞

Can't wait for it! This has been a blocker for C++ 20

JackBoosY commented on ports/shogun/eigen-3.4.patch:
```diff
@@ -0,0 +1,65 @@
+-			eigen_f.block(bl*n,0,n,1)=K*alpha->block(bl*n,0,n,1)*CMath::exp(log_scale*2.0);
+-			result+=alpha->block(bl*n,0,n,1).dot(eigen_f.block(bl*n,0,n,1))/2.0;
+-			eigen_f.block(bl*n,0,n,1)+=eigen_m;
++			eigen_f.segment(bl*n,n)=K*alpha->segment(bl*n,n)*CMath::exp(log_scale*2.0);
```
Are they upstream changes? If not, please make a PR about that to the upstream first.

aminya replied:
It's great to upstream changes, but this should not block this PR

NeighbourDYY replied:
https://github.com/microsoft/vcpkg/pull/19665#issuecomment-935064474
https://github.com/microsoft/vcpkg/pull/19665#issuecomment-990845905

Previous discussions even suggested giving up on shogun because of their own inactivity.

JackBoosY replied:
Fine.

spinicist commented:
🎉🥳 Thanks everyone who helped/approved/was involved with this
